book/src/using-bitfields.md

```diff
@@ -0,0 +1,196 @@
+# Using the Bitfield Types Generated by Bindgen
+
+For this discussion, we will use the following C type definitions and functions.
+```c
+typedef struct {
+    unsigned int a: 1;
+    unsigned int b: 1;
+    unsigned int c: 2;
+    
+} bitfield;
+
+// Create a default bitfield
+bitfield create_bitfield();
+
+// Print a bitfield
+void print_bitfield(bitfield bfield)
+```
+
+Bindgen creates a set of field getters and setters for interacting with the bitset. For example, 
+
+```rust,ignore
+    let mut bfield = unsafe { create_bitfield() };
+    
+    bfield.set_a(1);
+    println!("a set to {}", bfield.a());
+    bfield.set_b(1);
+    println!("b set to {}", bfield.b());
+    bfield.set_c(3);
+    println!("c set to {}", bfield.c());
+    unsafe { print_bitfield(bfield) };
+```
+
+will print out
+
+```text
+a set to 1
+b set to 1
+c set to 3
+bitfield: a:1, b:1, c:3
+```
+
+Overflowing a bitfield will result in the same behavior as in C/C++: the bitfield will be set to 0.
+
+```rust,ignore
+    let mut bfield = unsafe { create_bitfield() };
+    bfield.set_a(1);
+    bfield.set_b(1);
+    bfield.set_c(12);
+    println!("c set to {} due to overflow", bfield.c());
+    unsafe { print_bitfield(bfield) };
+```
+
+will print out
+
+```text
+c set to 0 due to overflow
+bitfield: a:1, b:1, c:0
+```
+
+To create a new bitfield in Rust, use mem::zeroed.
```

Comment:
Rather than use `std::mem::zeroed`, we should use the bitfield allocation unit constructors that `bindgen` emits: `new_bitfield_{1,2,...}`. About one of these is created for each set of contiguous bitfields within a struct.

```rust
let contains_bitfields = ContainsBitfields {
    _bitfield_1: ContainsBitfields::new_bitfield_1(/* a */ 1, /* b */ 2, /* c */ 3),
    ..Default::default()
};
```

Reply:
@fitgen with the currently generated code, the bitfield allocation unit constructors do not necessarily return a value of the same type as the bitfield allocation unit: see [this example](https://play.rust-lang.org/?gist=332bf8fcac476c5f723e393136adb80c&version=stable). What should the user do in this case?

Reply:
For now `transmute`, but really we should (eventually) fix that.

The problem is `transmute` is not a `const fn` yet, and we make these bitfield constructors `const fn`s when we can. Additionally, making `transmute` into a `const fn` requires Miri integration into rustc, and I'm not sure on the timeline for that.

We could do all kinds of things to work around this like:

* avoid generating the ctors when we need to `transmute` and its a `const fn`
* avoid making the ctor a `const fn` when we need to `transmute`
* skip generating any ctor at all when we need to `transmute`
* etc

Reply:
@fitzgen Is a fix along [these lines](https://play.rust-lang.org/?gist=f678edeea1a833e2051402416a18e7d7&version=stable) reasonable?

Reply:
@fitzgen During testing, I ran into an issue where Default isn't implemented for the bitfield. Is there a flag I need to set for that to be generated?